mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerTimezone.java

```diff
@@ -0,0 +1,165 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.hive;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.Collection;
+import java.util.List;
+import java.util.TimeZone;
+import org.apache.hadoop.hive.serde2.io.DateWritable;
+import org.apache.hadoop.hive.serde2.io.TimestampWritable;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.mr.TestHelper;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.types.Types;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.junit.runners.Parameterized.Parameter;
+import static org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class TestHiveIcebergStorageHandlerTimezone {
+
+  @Parameters(name = "timezone={0}")
+  public static Collection<Object[]> parameters() {
+    Collection<Object[]> testParams = ImmutableList.of(
+        new String[] {"America/New_York"},
+        new String[] {"Asia/Kolkata"},
+        new String[] {"UTC/Greenwich"}
+    );
+
+    return testParams;
+  }
+
+  private TestHiveShell shell;
+
+  private TestTables testTables;
+
+  private TimeZone originalTz = TimeZone.getDefault();
+
+  @Parameter(0)
+  public String timezoneString;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Before
+  public void before() throws IOException, IllegalAccessException {
+    TimeZone.setDefault(TimeZone.getTimeZone(timezoneString));
+
+    try {
+      // Magic to update cached stuff for Hive where the default timezone is used
+      Field formatterField = TimestampWritable.class.getDeclaredField("threadLocalDateFormat");
+      formatterField.setAccessible(true);
+      ThreadLocal<DateFormat> formatter = (ThreadLocal<DateFormat>) formatterField.get(null);
```

Comment:
Most Iceberg code uses `org.apache.iceberg.common.Dyn*` classes to handle reflection. Can we use that instead?

```
DynFields.StaticField<ThreadLocal<DateFormat>> dateFormat = DynFields.builder()
        .hiddenImpl(TimestampWritable.class, "threadLocalDateFormat")
        .defaultAlwaysNull() // it will return a StaticField which returns null instead of NoSuchFieldException on .get
        .buildStatic();
```

And `dateFormat.get()` to get the ThreadLocal

Reply:
Thanks for the suggestion!
Tidied up the code quite a bit.